hax/hax/motr/planner.py

```diff
@@ -0,0 +1,207 @@
+       threads.
+
+    2. The parallelism doesn't break semantics (some messages can be processed
+       in parallel to others, while some other not; the ConsumerThread's don't
+       need to worry about that).
+    """
+    def __init__(self):
+        self.state = State(next_group_id=0,
+                           active_commands=LinkedSet(),
+                           taken_commands=LinkedSet(),
+                           current_group_id=0,
+                           next_group_commands=LinkedSet(),
+                           is_shutdown=False)
+        self.backlog: Deque[BaseMessage] = deque()
+        self.b_lock = Condition()
+
+    def is_empty(self) -> bool:
+        """Checks whether the backlog is empty. Blocking call."""
+        with self.b_lock:
+            return not self.backlog
+
+    def add_command(self, command: BaseMessage) -> None:
+        LOG.log(TRACE, '[WP]Before add_command: %s', command)
+        with self.b_lock:
+            cmd = self._assign_group(command)
+            LOG.log(TRACE, '[WP]Cmd %s is added. Current state: %s', cmd,
+                    self.state)
+            self.backlog.append(cmd)
+            # Some threads may be waiting because of an empty backlog - let's
+            # notify them
+            self.b_lock.notifyAll()
+
+    def _create_poison(self) -> BaseMessage:
+        cmd = Die()
+        cmd.group = self.state.current_group_id
+        return cmd
+
+    def get_next_command(self) -> BaseMessage:
+        while True:
+            LOG.log(TRACE, '[WP]Trying to get new command')
+            with self.b_lock:
+                if self.state.is_shutdown:
+                    return self._create_poison()
+                if self.backlog:
+                    cmd = self.backlog.popleft()
+                    LOG.log(TRACE, '[WP]Cmd %s taken!', cmd)
+                    self.state.taken_commands.add(cmd)
+                    return cmd
+                LOG.log(TRACE, '[WP]Blocking thread: no commands in backlog')
+                self.b_lock.wait()
+
+    def shutdown(self):
+        with self.b_lock:
+            LOG.debug('WorkPlanner is shutting down')
+            self.state.is_shutdown = True
+            self.b_lock.notifyAll()
+
+    def ensure_allowed(self, command: BaseMessage) -> BaseMessage:
+        def is_current(cmd: BaseMessage, st: State) -> bool:
+            return cmd.group == st.current_group_id
+
+        while True:
+            with self.b_lock:
+                state = self.state
+                if state.is_shutdown:
+                    return self._create_poison()
+                if is_current(command, state):
+                    state.taken_commands.remove(command)
+                    state.active_commands.add(command)
+                    return command
+                LOG.log(TRACE, '[WP]Cmd %s not allowed for now.'
+                        ' Current state: %s', command, self.state)
+                self.b_lock.wait()
+
+    def _inc_group(self):
```

Comment:
- How many groups are expected to present at a given instance?
- Will the number of groups vary? If not then it would be good have an invariant that asserts the expected state of the planner at any given instance during its lifetime.

Reply:
That's a good point. After I wrote this I realized that the group index can grow too fast which is a problem namely in Python. It uses long integer maths just by default, so, for instance, if an integer number becomes greater than 4 bytes can hold, more memory will be used (no integer overflow will happen). And the fact is that we definitely don't need such huge numbers.

I'll update the PR to address that problem.

Reply:
Addressed, improved unit tests